piptools/_compat/pip_compat.py

```diff
@@ -94,7 +113,40 @@ def parse_requirements(
             file_link = FileLink(install_req.link.url)
             file_link._url = parsed_req.requirement
             install_req.link = file_link
-        yield copy_install_requirement(install_req)
+        install_req = copy_install_requirement(install_req)
+        if rewrite_comes_from is None:
+            pass
+        elif isinstance(rewrite_comes_from, str):
+            install_req.comes_from = rewrite_comes_from
+        else:
+            install_req.comes_from = rewrite_comes_from(install_req.comes_from)
```

Comment:
What if we flip these?
```suggestion
        if isinstance(rewrite_comes_from, str):
            install_req.comes_from = rewrite_comes_from
        elif rewrite_comes_from is not None:
            install_req.comes_from = rewrite_comes_from(install_req.comes_from)
```